{
 "cells": [
  {
   "cell_type": "markdown",
   "id": "160deb8e-353b-4b1b-85d3-2bee0a10a2f4",
   "metadata": {},
   "source": [
    "## Installing and importing packages and softwares"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "eee9dd68-949e-4d89-a75c-b660db9aa055",
   "metadata": {},
   "outputs": [],
   "source": [
    "import matplotlib as mpl\n",
    "import matplotlib.pyplot as plt\n",
    "from matplotlib.pyplot import rc_context\n",
    "from matplotlib import rcParams\n",
    "import matplotlib.colors as mcolors\n",
    "\n",
    "import numpy as np\n",
    "import distinctipy\n",
    "from distinctipy import colorblind\n",
    "import pandas as pd\n",
    "import warnings\n",
    "\n",
    "import anndata as ad\n",
    "import scanpy as sc\n",
    "import scipy\n",
    "\n",
    "import seaborn as sns\n",
    "\n",
    "import umap.umap_ as umap\n",
    "#import umap.plot_ as umap_plot\n",
    "\n",
    "from ipyfilechooser import FileChooser\n",
    "\n",
    "#For Tables:\n",
    "from IPython.display import display\n",
    "from IPython.display import Latex\n",
    "pd.set_option('display.max_columns', 500)\n",
    "from tqdm import tnrange, tqdm_notebook\n",
    "#%load_ext rpy2.ipython\n",
    "%reload_ext rpy2.ipython\n",
    "\n",
    "\n",
    "#Showing different spreads and min_dists - from https://scanpy.readthedocs.io/en/stable/tutorials/plotting/advanced.html\n",
    "from itertools import product"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "5bd403b1-6496-4cae-8f1f-67956b9efa23",
   "metadata": {},
   "source": [
    "## Import AnnData (adata) object that was converetd from a Seurat object made in Trailmaker"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "1466b765-f24c-47a3-bf83-59fc0fcddcbb",
   "metadata": {},
   "outputs": [],
   "source": [
    "adata = sc.read_h5ad(\"/Users/johnbriseno/Desktop/TMP_4_ZDrive/Data/snRNASeq/SAMap/h5ad/trailmaker_integrated_MWB_seurat_2_scanpy.h5ad\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "374d6a86-3e73-4a22-90c4-b7efb2386bcd",
   "metadata": {},
   "outputs": [],
   "source": [
    "adata"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "5f500b65-edb7-4330-a61d-fcaa952c28f6",
   "metadata": {},
   "source": [
    "## Downsample 10X nuclei and Parse nuclei to get 4000 for each sample\n",
    "Used example from here: https://github.com/scverse/scanpy/issues/987"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "502f3e3b-390b-4e9f-92fd-7c234cda80ef",
   "metadata": {},
   "outputs": [],
   "source": [
    "target_cells = 3450 #number nuclei for each sample\n",
    "cluster_key = \"samples\"\n",
    "\n",
    "grouped = adata.obs.groupby(cluster_key)\n",
    "downsampled_indices = []\n",
    "\n",
    "for _, group in grouped:\n",
    "    if len(group) > target_cells:\n",
    "        downsampled_indices.extend(group.sample(target_cells).index)\n",
    "    else:\n",
    "        downsampled_indices.extend(group.index)\n",
    "\n",
    "adata_downsampled = adata[downsampled_indices]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "60da2249-b3df-4b83-913b-1758826a5250",
   "metadata": {},
   "outputs": [],
   "source": [
    "adata_downsampled"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "9e207e68-e994-419d-ad71-8152ab09e197",
   "metadata": {},
   "source": [
    "## Looking at Trailmaker (Seurat) UMAP embeddings and changing some metadata"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "432080f1-3f4b-4242-9bab-d2b172c26818",
   "metadata": {},
   "outputs": [],
   "source": [
    "sc.pl.umap(adata_downsampled, \n",
    "           color = ['seurat_clusters',\n",
    "                    'samples'\n",
    "                   ], \n",
    "           legend_loc = 'on data', legend_fontsize = 'xx-small', \n",
    "           ncols = 2,\n",
    "           #save = '_trailmaker_seurat_Male_WBs_clusters2_UMAPs.svg'\n",
    "          )"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "02e572b2-eb01-4d5d-96bc-9f3ec7785e57",
   "metadata": {},
   "outputs": [],
   "source": [
    "#Changing sample color\n",
    "sc.pl.umap(adata_downsampled, \n",
    "           color = ['samples'], \n",
    "           legend_loc = 'on data', legend_fontsize = 'xx-small', \n",
    "           ncols = 2,\n",
    "           palette={\"MZ7\": \"tab:blue\", \"MA7\": \"tab:purple\"},\n",
    "           #save = '_Male_WBs_clusters2_UMAPs.svg'\n",
    "          )"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "f9396682-f167-4e57-aae7-e98f5c60b30b",
   "metadata": {},
   "source": [
    "## Rerun Analysis in Scanpy\n",
    "With help from this tutorial: https://nbisweden.github.io/workshop-scRNAseq/labs/scanpy/scanpy_07_trajectory.html"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "1b544616-061b-4770-89e9-e10815c0f44c",
   "metadata": {},
   "outputs": [],
   "source": [
    "#first, store the old umap with a new name so it is not overwritten\n",
    "#adata_downsampled.obsm['X_umap_old'] = adata_downsampled.obsm['X_umap'] #-> only run this once\n",
    "sc.pp.neighbors(adata_downsampled, \n",
    "                n_pcs = 40, #same PCs as trailamker\n",
    "                n_neighbors = 20, #what was trialmaker run with?\n",
    "                use_rep=\"X_harmony\") #we used 4000 HVGs\n",
    "sc.tl.umap(adata_downsampled, min_dist=0.4, spread=3)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "73730086-5821-46c9-a1c0-74f6aaa99624",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Redo clustering as well\n",
    "#sc.tl.leiden(adata_downsampled, key_added = \"leiden_0.8\", resolution = 0.8) \n",
    "#sc.tl.leiden(adata_downsampled, key_added = \"leiden_1.0\", resolution = 1.0)\n",
    "#sc.tl.leiden(adata_downsampled, key_added = \"leiden_1.2\", resolution = 1.2)\n",
    "#sc.tl.leiden(adata_downsampled, key_added = \"leiden_1.4\", resolution = 1.4)\n",
    "\n",
    "sc.pl.umap(adata_downsampled, color = ['seurat_clusters', 'leiden_0.8', 'leiden_1.0', 'leiden_1.2', 'leiden_1.4'],\n",
    "           legend_loc = 'on data', legend_fontsize = 'xx-small', ncols =1,\n",
    "           #save=\"_scanpy_clusters_various_resolutons_n_neighbor_20.png\"\n",
    "          )"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "b8b93b89-9bf3-4d45-89e3-1b5c72b58e09",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "markdown",
   "id": "9fd831ec-aa70-4dc7-8329-01417c46e921",
   "metadata": {},
   "source": [
    "#### saving clustering so we can just run from here next time"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "e64e36b5-e898-4888-9266-adacd510d547",
   "metadata": {},
   "outputs": [],
   "source": [
    "adata_downsampled.write(\"/Users/johnbriseno/Desktop/TMP_4_ZDrive/Data/snRNASeq/SAMap/h5ad/adata_downsampled_trailmaker_integrated_MWB_seurat_2_scanpy.h5ad\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "c58d41eb-461e-4c7a-8673-5c64dc47c2df",
   "metadata": {},
   "outputs": [],
   "source": [
    "#reading back in h5ad\n",
    "adata_downsampled = sc.read_h5ad(\"/Users/johnbriseno/Desktop/TMP_4_ZDrive/Data/snRNASeq/SAMap/h5ad/adata_downsampled_trailmaker_integrated_MWB_seurat_2_scanpy.h5ad\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "92bdd038-e014-4a01-901f-911902858d55",
   "metadata": {},
   "outputs": [],
   "source": [
    "# plot new scanpy clusters onto the seurat embedding:\n",
    "sc.pl.embedding(adata_downsampled, basis='X_umap_old', color = ['seurat_clusters','leiden_0.8', 'leiden_1.0', 'leiden_1.2', 'leiden_1.4'],\n",
    "                legend_loc = 'on data', legend_fontsize = 'xx-small', ncols =1,\n",
    "               save=\"_scanpy_clusters__n.neighbor20_on_trailmaker_Serurat_embeddings.png\"\n",
    "               )"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "b09ec326-492c-4bc1-a093-c406af7c1ba7",
   "metadata": {},
   "outputs": [],
   "source": [
    "sc.tl.umap(adata_downsampled, min_dist=0.4, spread=3, n_components=3) #making these 3d off the bat\n",
    "sc.pl.umap(adata_downsampled, color = 'seurat_clusters', projection='3d', save=\"3dumap_seurat_clusters.png\")\n",
    "sc.pl.umap(adata_downsampled, color = 'leiden_0.8', projection='3d', save=\"3dumap_leiden_0.8.png\")\n",
    "sc.pl.umap(adata_downsampled, color = 'leiden_1.0', projection='3d', save=\"3dumap_leiden_1.0.png\")\n",
    "sc.pl.umap(adata_downsampled, color = 'leiden_1.2', projection='3d', save=\"3dumap_leiden_1.2.png\")\n",
    "sc.pl.umap(adata_downsampled, color = 'leiden_1.4', projection='3d', save=\"3dumap_leiden_1.4.png\")"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "a96f5119-92d3-4580-8ee0-51bdf9443cbf",
   "metadata": {},
   "source": [
    "## Checkout clusters from sample\n",
    "with frequency plot example from here: https://www.biostars.org/p/9575906/"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "5b74c126-8452-4953-87d2-f8dfa35664de",
   "metadata": {},
   "outputs": [],
   "source": [
    "color = ['purple', 'blue']\n",
    "\n",
    "tmp = pd.crosstab(adata_downsampled.obs['seurat_clusters'],adata.obs['samples'], normalize='index')\n",
    "tmp.plot.bar(stacked=True).legend(loc='upper right')\n",
    "tmp = pd.crosstab(adata_downsampled.obs['leiden_0.8'],adata.obs['samples'], normalize='index')\n",
    "tmp.plot.bar(stacked=True).legend(loc='upper right')\n",
    "tmp = pd.crosstab(adata_downsampled.obs['leiden_1.0'],adata.obs['samples'], normalize='index')\n",
    "tmp.plot.bar(stacked=True).legend(loc='upper right')\n",
    "tmp = pd.crosstab(adata_downsampled.obs['leiden_1.2'],adata.obs['samples'], normalize='index')\n",
    "tmp.plot.bar(stacked=True).legend(loc='upper right')\n",
    "tmp = pd.crosstab(adata_downsampled.obs['leiden_1.4'],adata.obs['samples'], normalize='index')\n",
    "tmp.plot.bar(stacked=True).legend(loc='upper right')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "fe45558f-478b-4192-9022-ee508e2844b1",
   "metadata": {},
   "outputs": [],
   "source": [
    "#getting colors -> https://stackoverflow.com/questions/11927715/how-to-give-a-pandas-matplotlib-bar-graph-custom-colors\n",
    "#saving fig -> https://matplotlib.org/stable/api/_as_gen/matplotlib.pyplot.savefig.html\n",
    "#getting counts, not relative percentages -> get rid of normalize='columns' from pd.crosstab(y,x, nomralize = 'columns')\n",
    "\n",
    "\n",
    "tmp = pd.crosstab(adata_downsampled.obs['seurat_clusters'],adata_downsampled.obs['samples']).T.plot(kind='bar', stacked=True, color=adata_downsampled.uns['seurat_clusters_colors'])\n",
    "tmp.legend(title='seurat_clusters', bbox_to_anchor=(1.26, 1.02),loc='upper right',labelcolor=adata_downsampled.uns['seurat_clusters_colors'])\n",
    "plt.savefig('/Users/johnbriseno/Desktop/M_WBs_Jupyter/figures/seurat_clusters_freq_plot.png', \n",
    "            bbox_inches='tight', pad_inches=0.1,\n",
    "            dpi=400, format='png')\n",
    "\n",
    "tmp = pd.crosstab(adata_downsampled.obs['leiden_0.8'],adata_downsampled.obs['samples']).T.plot(kind='bar', stacked=True, color=adata_downsampled.uns['leiden_0.8_colors'])\n",
    "tmp.legend(title='leiden_0.8', bbox_to_anchor=(1.26, 1.02),loc='upper right',labelcolor=adata_downsampled.uns['leiden_0.8_colors'])\n",
    "plt.savefig('/Users/johnbriseno/Desktop/M_WBs_Jupyter/figures/leiden_0.8_freq_plot.png', \n",
    "            bbox_inches='tight', pad_inches=0.1,\n",
    "            dpi=400, format='png')\n",
    "\n",
    "tmp = pd.crosstab(adata_downsampled.obs['leiden_1.0'],adata_downsampled.obs['samples']).T.plot(kind='bar', stacked=True, color=adata_downsampled.uns['leiden_1.0_colors'])\n",
    "tmp.legend(title='leiden_1.0', bbox_to_anchor=(1.26, 1.02),loc='upper right',labelcolor=adata_downsampled.uns['leiden_1.0_colors'])\n",
    "plt.savefig('/Users/johnbriseno/Desktop/M_WBs_Jupyter/figures/leiden_1.0_freq_plot.png', \n",
    "            bbox_inches='tight', pad_inches=0.1,\n",
    "            dpi=400, format='png')\n",
    "\n",
    "tmp = pd.crosstab(adata_downsampled.obs['leiden_1.2'],adata_downsampled.obs['samples']).T.plot(kind='bar', stacked=True, color=adata_downsampled.uns['leiden_1.2_colors'])\n",
    "tmp.legend(title='leiden_1.2', bbox_to_anchor=(1.26, 1.02),loc='upper right',labelcolor=adata_downsampled.uns['leiden_1.2_colors'])\n",
    "plt.savefig('/Users/johnbriseno/Desktop/M_WBs_Jupyter/figures/leiden_1.2_freq_plot.png', \n",
    "            bbox_inches='tight', pad_inches=0.1,\n",
    "            dpi=400, format='png')\n",
    "\n",
    "tmp = pd.crosstab(adata_downsampled.obs['leiden_1.4'],adata_downsampled.obs['samples']).T.plot(kind='bar', stacked=True, color=adata_downsampled.uns['leiden_1.4_colors'])\n",
    "tmp.legend(title='leiden_1.4', bbox_to_anchor=(1.26, 1.02),loc='upper right',labelcolor=adata_downsampled.uns['leiden_1.4_colors'])\n",
    "plt.savefig('/Users/johnbriseno/Desktop/M_WBs_Jupyter/figures/leiden_1.4_freq_plot.png', \n",
    "            bbox_inches='tight', pad_inches=0.1,\n",
    "            dpi=400, format='png')\n",
    "\n",
    "#do this but have raw count? - normalize='columns' from pd.crosstab(y,x, nomralize = 'columns')\n",
    "#need to fix legends "
   ]
  },
  {
   "cell_type": "markdown",
   "id": "5c905c95-ecd4-482d-a113-bca8065ec991",
   "metadata": {},
   "source": [
    "## Ranking genes to find cluster-specific markers and across resolutions\n",
    "https://scanpy-tutorials.readthedocs.io/en/multiomics/visualizing-marker-genes.html#Visualization-of-marker-genes\n",
    "\n",
    "https://github.com/scverse/scanpy/issues/748 <- good explanation of marker genes, \n",
    "such that \"tells you which genes characterize a cluster, but doesn't necessarily tell you \n",
    "which genes contributed most to the global split of clusters that was generated\n",
    "\n",
    "With these, can do marker gene visualization downstream.\n",
    "\n",
    "Let's just get unique genes of interest too (must filter out adata which is just a df)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "d851b785-30b4-4d52-bd06-f62ec0e38822",
   "metadata": {},
   "outputs": [],
   "source": [
    "\n",
    "#must call all rankings by their own names, otherwise just rewriting a bunch of stuff\n",
    "\n",
    "sc.tl.rank_genes_groups(adata_downsampled, \n",
    "                        groupby=\"seurat_clusters\",\n",
    "                        method=\"wilcoxon\",\n",
    "                        corr_method=\"benjamini-hochberg\", \n",
    "                        key_added=\"wilcoxon_seurat_clusters\")\n",
    "#Apply more stringent filter\n",
    "sc.tl.filter_rank_genes_groups(adata_downsampled,\n",
    "                               groupby=\"seurat_clusters\",\n",
    "                               key='wilcoxon_seurat_clusters',\n",
    "                               min_fold_change=1,\n",
    "                               key_added='wilcoxon_filtered_seurat_clusters')\n",
    "\n",
    "\n",
    "sc.tl.rank_genes_groups(adata_downsampled, \n",
    "                        groupby=\"leiden_0.8\",\n",
    "                        method=\"wilcoxon\",\n",
    "                        corr_method=\"benjamini-hochberg\", \n",
    "                        key_added=\"wilcoxon_leiden_0.8\")\n",
    "#Apply more stringent filter\n",
    "sc.tl.filter_rank_genes_groups(adata_downsampled,\n",
    "                               groupby=\"leiden_0.8\",\n",
    "                               key='wilcoxon_leiden_0.8',\n",
    "                               min_fold_change=1,\n",
    "                               key_added='wilcoxon_filtered_leiden_0.8')\n",
    "\n",
    "\n",
    "sc.tl.rank_genes_groups(adata_downsampled, \n",
    "                        groupby=\"leiden_1.0\",\n",
    "                        method=\"wilcoxon\",\n",
    "                        corr_method=\"benjamini-hochberg\", \n",
    "                        key_added=\"wilcoxon_leiden_1.0\")\n",
    "#Apply more stringent filter\n",
    "sc.tl.filter_rank_genes_groups(adata_downsampled,\n",
    "                               groupby=\"leiden_1.0\",\n",
    "                               key='wilcoxon_leiden_1.0',\n",
    "                               min_fold_change=1,\n",
    "                               key_added='wilcoxon_filtered_leiden_1.0')\n",
    "\n",
    "\n",
    "sc.tl.rank_genes_groups(adata_downsampled, \n",
    "                        groupby=\"leiden_1.2\",\n",
    "                        method=\"wilcoxon\",\n",
    "                        corr_method=\"benjamini-hochberg\", \n",
    "                        key_added=\"wilcoxon_leiden_1.2\")\n",
    "#Apply more stringent filter\n",
    "sc.tl.filter_rank_genes_groups(adata_downsampled,\n",
    "                               groupby=\"leiden_1.2\",\n",
    "                               key='wilcoxon_leiden_1.2',\n",
    "                               min_fold_change=1,\n",
    "                               key_added='wilcoxon_filtered_leiden_1.2')\n",
    "\n",
    "sc.tl.rank_genes_groups(adata_downsampled, \n",
    "                        groupby=\"leiden_1.4\",\n",
    "                        method=\"wilcoxon\",\n",
    "                        corr_method=\"benjamini-hochberg\", \n",
    "                        key_added=\"wilcoxon_leiden_1.4\")\n",
    "#Apply more stringent filter\n",
    "sc.tl.filter_rank_genes_groups(adata_downsampled,\n",
    "                               groupby=\"leiden_1.4\",\n",
    "                               key='wilcoxon_leiden_1.4',\n",
    "                               min_fold_change=1,\n",
    "                               key_added='wilcoxon_filtered_leiden_1.4')\n",
    "\n",
    "\n",
    "#annot"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "741bf1ce-3441-42f2-910b-7aec22e1ae41",
   "metadata": {},
   "outputs": [],
   "source": [
    "#running dendrogram to cluster clusters based on similarity \n",
    "sc.tl.dendrogram(adata_downsampled, groupby=\"seurat_clusters\")\n",
    "sc.tl.dendrogram(adata_downsampled, groupby=\"leiden_0.8\")\n",
    "sc.tl.dendrogram(adata_downsampled, groupby=\"leiden_1.0\")\n",
    "sc.tl.dendrogram(adata_downsampled, groupby=\"leiden_1.2\")\n",
    "sc.tl.dendrogram(adata_downsampled, groupby=\"leiden_1.4\")\n",
    "\n",
    "#checkout heatmaps from diff clusterings with 'rank_genes_groups'\n",
    "sc.pl.rank_genes_groups_heatmap(\n",
    "    adata_downsampled,\n",
    "    groupby='seurat_clusters',\n",
    "    n_genes=3,\n",
    "    show_gene_labels=True,\n",
    "    cmap=\"viridis\",\n",
    "    key='wilcoxon_seurat_clusters',\n",
    "    swap_axes=True,\n",
    "    save='_wilcoxon_seurat_clusters.png'\n",
    ")\n",
    "sc.pl.rank_genes_groups_heatmap(\n",
    "    adata_downsampled,\n",
    "    groupby='leiden_0.8',\n",
    "    n_genes=5,\n",
    "    show_gene_labels=True,\n",
    "    cmap=\"viridis\",\n",
    "    key='wilcoxon_leiden_0.8',\n",
    "    swap_axes=True,\n",
    "    save='_wilcoxon_leiden_0.8.png'\n",
    ")\n",
    "sc.pl.rank_genes_groups_heatmap(\n",
    "    adata_downsampled,\n",
    "    groupby='leiden_1.0',\n",
    "    n_genes=5,\n",
    "    show_gene_labels=True,\n",
    "    cmap=\"viridis\",\n",
    "    key='wilcoxon_leiden_1.0',\n",
    "    swap_axes=True,\n",
    "    save='_wilcoxon_leiden_1.0.png',\n",
    ")\n",
    "sc.pl.rank_genes_groups_heatmap(\n",
    "    adata_downsampled,\n",
    "    groupby='leiden_1.2',\n",
    "    n_genes=5,\n",
    "    show_gene_labels=True,\n",
    "    cmap=\"viridis\",\n",
    "    key='wilcoxon_leiden_1.2',\n",
    "    swap_axes=True,\n",
    "    save='_wilcoxon_leiden_1.2.png',\n",
    ")\n",
    "sc.pl.rank_genes_groups_heatmap(\n",
    "    adata_downsampled,\n",
    "    groupby='leiden_1.4',\n",
    "    n_genes=5,\n",
    "    show_gene_labels=True,\n",
    "    cmap=\"viridis\",\n",
    "    key='wilcoxon_leiden_1.4',\n",
    "    swap_axes=True,\n",
    "    save='_wilcoxon_leiden_1.4.png',\n",
    ")\n",
    "#annot"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "dd1370f2-e687-4885-b005-606b5aef9a15",
   "metadata": {},
   "outputs": [],
   "source": [
    "#checkout heatmaps from diff clusterings with 'rank_genes_groups'\n",
    "sc.pl.rank_genes_groups_heatmap(\n",
    "    adata_downsampled,\n",
    "    groupby='seurat_clusters',\n",
    "    n_genes=3,\n",
    "    show_gene_labels=True,\n",
    "    cmap=\"viridis\",\n",
    "    key='wilcoxon_filtered_seurat_clusters',\n",
    "    swap_axes=True,\n",
    "    save='_wilcoxon_filtered_seurat_clusters.png'\n",
    ")\n",
    "sc.pl.rank_genes_groups_heatmap(\n",
    "    adata_downsampled,\n",
    "    groupby='leiden_0.8',\n",
    "    n_genes=5,\n",
    "    show_gene_labels=True,\n",
    "    cmap=\"viridis\",\n",
    "    key='wilcoxon_filtered_leiden_0.8',\n",
    "    swap_axes=True,\n",
    "    save='_wilcoxon_filtered_leiden_0.8.png'\n",
    ")\n",
    "sc.pl.rank_genes_groups_heatmap(\n",
    "    adata_downsampled,\n",
    "    groupby='leiden_1.0',\n",
    "    n_genes=5,\n",
    "    show_gene_labels=True,\n",
    "    cmap=\"viridis\",\n",
    "    key='wilcoxon_filtered_leiden_1.0',\n",
    "    swap_axes=True,\n",
    "    save='_wilcoxon_filtered_leiden_1.0.png',\n",
    ")\n",
    "sc.pl.rank_genes_groups_heatmap(\n",
    "    adata_downsampled,\n",
    "    groupby='leiden_1.2',\n",
    "    n_genes=5,\n",
    "    show_gene_labels=True,\n",
    "    cmap=\"viridis\",\n",
    "    key='wilcoxon_filtered_leiden_1.2',\n",
    "    swap_axes=True,\n",
    "    save='_wilcoxon_filtered_leiden_1.2.png',\n",
    ")\n",
    "sc.pl.rank_genes_groups_heatmap(\n",
    "    adata_downsampled,\n",
    "    groupby='leiden_1.4',\n",
    "    n_genes=5,\n",
    "    show_gene_labels=True,\n",
    "    cmap=\"viridis\",\n",
    "    key='wilcoxon_filtered_leiden_1.4',\n",
    "    swap_axes=True,\n",
    "    save='_wilcoxon_filtered_leiden_1.4.png',\n",
    ")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "d934172b-d1cd-4cb9-a58c-75bfc51ff232",
   "metadata": {},
   "outputs": [],
   "source": [
    "#getting just unique markers per cluster\n",
    "#code from google ai agh \n",
    "\n",
    "#basically extract adata as dataframe and filter for just genes in cluster\n",
    "\n",
    "###Step 1: Run sc.tl.rank_genes_groups() and name each key independtly (did up top)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "44a7b06e-e8bf-4b01-878b-3df584f48fdb",
   "metadata": {},
   "outputs": [],
   "source": [
    "###Step 2: Extract the results into a pandas DataFrame\n",
    "results_df = sc.get.rank_genes_groups_df(adata_downsampled, group=None, key='wilcoxon_leiden_1.0')\n",
    "#group=None: This parameter tells the function to get the results for all groups, which is what you need to filter for uniqueness.\n",
    "#Must specify key since there is no 'rank_genes_groups' default - each DE filter is called by its resolution"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "7305f543-2e51-4239-a9a0-dca1d241113e",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Define a function to get unique markers\n",
    "def get_unique_markers(results_df, n_genes):\n",
    "    all_genes = []\n",
    "    unique_markers = {}\n",
    "    \n",
    "    # Get the list of clusters from the DataFrame\n",
    "    clusters = results_df['group'].unique()\n",
    "    \n",
    "    for cluster in clusters:\n",
    "        # Get top N genes for the current cluster\n",
    "        cluster_genes = results_df[results_df['group'] == cluster].head(n_genes)['names'].tolist()\n",
    "        \n",
    "        # Keep genes not yet seen in previous clusters\n",
    "        unique_genes = [gene for gene in cluster_genes if gene not in all_genes]\n",
    "        unique_markers[cluster] = unique_genes\n",
    "        \n",
    "        # Add the unique genes to the master list\n",
    "        all_genes.extend(unique_genes)\n",
    "        \n",
    "    return unique_markers"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "ae2b7c26-7965-4395-8faf-1e6fd090fc97",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Get the top 5 unique markers for each cluster\n",
    "n_genes_per_cluster = 5\n",
    "unique_markers_dict = get_unique_markers(results_df, n_genes_per_cluster)\n",
    "\n",
    "# Print the resulting dictionary of unique markers\n",
    "for cluster, genes in unique_markers_dict.items():\n",
    "    print(f\"Cluster {cluster}: {genes}\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "0b8168c9-83e0-4562-a336-6559a2410595",
   "metadata": {},
   "outputs": [],
   "source": [
    "sc.pl.dotplot(adata_downsampled, \n",
    "              #unique_markers_dict, \n",
    "              groupby='leiden_1.0',\n",
    "              dendrogram=True,\n",
    "              swap_axes=True,\n",
    "              #save=\"_unique_markers_leiden_1.0_dotplot.svg\"\n",
    "             )"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "65332a42-b896-43ac-9576-1a013037d248",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "4cd048d2-6188-4073-a432-05357b96f81c",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "0536d954-18cc-433b-9404-f5d27d3f475f",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "c329bc2c-a4fa-4afc-8437-7621bc2592b6",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "markdown",
   "id": "d7b9cd7a-e7ed-4821-b940-f9dea1d5730b",
   "metadata": {
    "jp-MarkdownHeadingCollapsed": true
   },
   "source": [
    "## Also try with really low resolutions to checkout \n",
    "Resolution is largely arbitrary but want to make sure we have strong cluster-specific markers"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "fa55c9af-b1ab-4c9d-9499-846c2bcbc273",
   "metadata": {},
   "outputs": [],
   "source": [
    "#also try with really low resolutions to checkout \n",
    "sc.tl.leiden(adata_downsampled, key_added = \"leiden_0.02\", resolution = 0.02) # default resolution in 0.8\n",
    "sc.tl.leiden(adata_downsampled, key_added = \"leiden_0.05\", resolution = 0.05) # default resolution in 1.0\n",
    "sc.tl.leiden(adata_downsampled, key_added = \"leiden_0.075\", resolution = 0.075) # default resolution in 1.2\n",
    "sc.tl.leiden(adata_downsampled, key_added = \"leiden_0.1\", resolution = 0.1) # default resolution in 1.4\n",
    "sc.tl.leiden(adata_downsampled, key_added = \"leiden_0.5\", resolution = 0.5) # default resolution in 1.4"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "68ca80dc-4b55-409e-896b-4d8f25dcba75",
   "metadata": {},
   "outputs": [],
   "source": [
    "sc.pl.umap(adata_downsampled, color = ['leiden_0.02', 'leiden_0.05', 'leiden_0.075', 'leiden_0.1', 'leiden_0.5'],legend_loc = 'on data', legend_fontsize = 'xx-small', ncols =2)\n",
    "#looks like only 0.5 looks reasonable"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "f296cb19-838c-4704-a34d-6412bff61def",
   "metadata": {},
   "outputs": [],
   "source": [
    "#frequency plots\n",
    "tmp = pd.crosstab(adata_downsampled.obs['leiden_0.5'],adata.obs['samples'], normalize='index')\n",
    "tmp.plot.bar(stacked=True).legend(loc='upper right')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "292bb414-498c-43ea-8c50-0c5fd16fd7f2",
   "metadata": {},
   "outputs": [],
   "source": [
    "#markeger genes\n",
    "sc.tl.rank_genes_groups(adata_downsampled, \n",
    "                        groupby=\"leiden_0.5\",\n",
    "                        groups='all',\n",
    "                        reference='rest',\n",
    "                        method=\"wilcoxon\",\n",
    "                        corr_method=\"benjamini-hochberg\", key_added=\"rank_genes_groups\")\n",
    "#Apply more stringent filter\n",
    "sc.tl.filter_rank_genes_groups(adata_downsampled,\n",
    "                               groupby=\"leiden_0.5\",\n",
    "                               min_in_group_fraction=0.25,\n",
    "                               max_out_group_fraction=0.5,\n",
    "                               min_fold_change=1, key_added='rank_genes_groups_filtered')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "d3fcdfdf-5444-4405-ae80-79641804584f",
   "metadata": {},
   "outputs": [],
   "source": [
    "sc.pl.rank_genes_groups_heatmap(\n",
    "    adata_downsampled,\n",
    "    groupby='leiden_0.5',\n",
    "    n_genes=10,\n",
    "    show_gene_labels=True,\n",
    "    cmap=\"viridis\",\n",
    "    key='rank_genes_groups_filtered',\n",
    "    swap_axes=True\n",
    ")\n",
    "sc.pl.rank_genes_groups_stacked_violin(\n",
    "    adata_downsampled,\n",
    "    groupby='leiden_0.5',\n",
    "    n_genes=10,\n",
    "    cmap=\"viridis\",\n",
    "    key='rank_genes_groups_filtered',\n",
    ")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "c9e151dc-5666-4433-8e74-41c31e5e6856",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "markdown",
   "id": "f38f1966-a5a2-443d-be4f-0be28ec248c3",
   "metadata": {},
   "source": [
    "#### Now pick a resolution to run on downstream analyses"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "410e309f-b9c6-43b1-99b0-7039ac697fca",
   "metadata": {},
   "source": [
    "## Visualizing known marker genes"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "ce2a4591-b103-4c67-9948-829d6859a003",
   "metadata": {},
   "outputs": [],
   "source": [
    "#GOI plots\n",
    "#Setting gene directories\n",
    "markers_genes_dict = {\n",
    "    'Putative Hematopoietic Markers':['g29365', #HHEX\n",
    "                                      'g21408', #MARCO\n",
    "                                      'g24607', #ssc4D\n",
    "                                      'g25995' #hephl1\n",
    "                                     ],\n",
    "    'WB Homolog Single-Cell Markers': ['g776', #SVEP1\n",
    "                                       'g19257', #HSPG2\n",
    "                                       'g13641', #PCNA\n",
    "                                       'g8344' #gata3\n",
    "                                      ],\n",
    "    'Hemocyte Markers': ['cluster-2407', #EsPGRP5\n",
    "                         'g2715', #EsTLR\n",
    "                         'g4518', #Esgalectin2\n",
    "                         'g18324' #myd88\n",
    "                        ]\n",
    "}\n",
    "type(markers_genes_dict)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "302ef89a-5be4-43de-82ac-5608f7149b07",
   "metadata": {},
   "outputs": [],
   "source": [
    "markers_list = [\n",
    "    #wb bulk markers\n",
    "    'g29365', #HHEX\n",
    "    'g21408', #MARCO\n",
    "    'g24607', #ssc4D\n",
    "    'g25995', #hephl1\n",
    "\n",
    "    #wb sc markers\n",
    "    'g776', #SVEP1\n",
    "    'g19257', #HSPG2\n",
    "    'g13641', #PCNA\n",
    "    'g8344', #gata3\n",
    "    \n",
    "    #hct markers\n",
    "    'cluster-2407', #EsPGRP5\n",
    "    'g2715', #EsTLR\n",
    "    'g4518', #Esgalectin2\n",
    "    'g18324' #myd88\n",
    "          ]\n",
    "type(markers_list)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "00becef3-075d-4dda-b11d-ffba10113f66",
   "metadata": {},
   "outputs": [],
   "source": [
    "#gotta remake umap as 2d\n",
    "#sc.tl.umap(adata_downsampled, min_dist=0.4, spread=3)\n",
    "\n",
    "sc.pl.umap(adata_downsampled, color = markers_list, \n",
    "           ncols= 2, cmap=sns.cubehelix_palette(dark=0, light=.9, as_cmap=True))\n",
    "#use_raw = False -> raw genes (adata.raw.var_names) are those ~39K genes not in the 2000 HVGs used to integrate the data (which are in adata.var_names)\n",
    "#lowly expressed genes (ie HHEX) are not HVGs and thus in raw\n",
    "#cmap from here:https://github.com/scverse/scanpy/issues/1550"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "5291ff7e-bfb4-4ff5-9af3-fdc73b3cdfa7",
   "metadata": {},
   "outputs": [],
   "source": [
    "#From stakced violin tutorial:\n",
    "#https://scanpy.readthedocs.io/en/stable/generated/scanpy.pl.stacked_violin.html\n",
    "\n",
    "#How to save these matplotlib figs with add_totals():\n",
    "#https://stackoverflow.com/questions/78463183/problem-exporting-images-with-matplot-lib\n",
    "\n",
    "sc.tl.dendrogram(adata_downsampled, groupby = 'leiden_1.0')\n",
    "\n",
    "fig, (ax1, ax2) = plt.subplots(1, 2, figsize=(15, 10), gridspec_kw={\"wspace\": 0.2})\n",
    "\n",
    "ax1_dict = sc.pl.dotplot(\n",
    "    adata_downsampled, markers_genes_dict, groupby=\"leiden_1.0\", ax=ax1, var_group_rotation=45,#, swap_axes= True,\n",
    "    show=False, \n",
    "    dendrogram = True)\n",
    "\n",
    "vp = sc.pl.stacked_violin(\n",
    "    adata_downsampled, markers_genes_dict, groupby=\"leiden_1.0\", ax=ax2, var_group_rotation=45, swap_axes= True,\n",
    "    show=True, \n",
    "    return_fig = True)\n",
    "vp.add_totals(sort = 'ascending', color= adata_downsampled.uns['leiden_1.0_colors']).style(ylim=(0,5)).show()\n",
    "\n",
    "\n",
    "#fig.savefig(\"/Users/johnbriseno/Desktop/M_WBs_Jupyter/figures/WB_GOIs.svg\", format=\"svg\")\n",
    "fig.show()\n",
    "\n",
    "\n",
    "#annot\n",
    "#seurat_clusters"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "a2022f1a-0a02-4eb7-a523-edb06aa3f317",
   "metadata": {},
   "source": [
    "#### Another GOI dictionary for model figure"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "17f91320-046e-4f69-ab13-29e065c3986a",
   "metadata": {},
   "outputs": [],
   "source": [
    "#GOI plots\n",
    "#Setting gene directories\n",
    "funct_marker_genes_dict = {\n",
    "    'MTC Membrane Transporters':['g15814', #MTFR1\n",
    "                                 'g3979', #SLC25A31\n",
    "                                 'g28362' #VDAC2\n",
    "                                ],\n",
    "    'Transcription Factors':['g29365', #HHEX\n",
    "                             'g8344', #GATA3\n",
    "                             'g15292' #SOX7\n",
    "                            ],\n",
    "    'RNA Transport & Stability':[#'g16209', #HNRNPA2B1\n",
    "                                 'g16208', #HNRNPA2B1\n",
    "                                 'g17243', #HNRNPA2B1\n",
    "                                 'g13639' #HNRNPH1\n",
    "                                ],\n",
    "    'Actin-Mediated Cytoplasmic Bridge':['g2950', #Tm1\n",
    "                                        'g18044', #FSCN1\n",
    "                                        'g10567' #COTL1\n",
    "                                        ],\n",
    "    'Ras GTPase Signaling':['g14463', #ARHGDIA\n",
    "                           'g18872', #RHOA\n",
    "                           'g26916' #RABGAP1\n",
    "                           ],\n",
    "    'Ion Binding & Transport':['g25995', #HEPHL1\n",
    "                              'g29902', #FTH1\n",
    "                              'g27264' #SPTAN1\n",
    "                              ],\n",
    "    'Pattern Recognition Receptors':['cluster-2407', #EsPGRP5\n",
    "                                     'g21408', #MARCO\n",
    "                                     'g24607' #SSC4D\n",
    "                                    ]\n",
    "}\n",
    "type(funct_marker_genes_dict)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "85a15f1d-9fe0-4751-83fd-a13ef42a7aa1",
   "metadata": {},
   "outputs": [],
   "source": [
    "sc.pl.stacked_violin(adata_downsampled, \n",
    "                     funct_marker_genes_dict, groupby=\"leiden_1.0\", \n",
    "                     dendrogram = True,\n",
    "                     var_group_rotation=45, swap_axes= True,\n",
    "                     cmap='viridis',\n",
    "                    save = '_violin_model_GOIs.svg')"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "46a288c8-4e96-44f9-8be6-2628a79f6927",
   "metadata": {},
   "source": [
    "## Run PAGA  on these new nuclei embeddings and resolution\n",
    "Use the clusters from leiden clustering run PAGA. First we create the graph and initialize the positions using the umap."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "ccd05595-345e-44ac-b878-c6e0aacfec53",
   "metadata": {},
   "outputs": [],
   "source": [
    "# use the umap to initialize the graph layout.\n",
    "sc.tl.draw_graph(adata_downsampled, init_pos='X_umap')\n",
    "\n",
    "#use force atlas 2 to spatialize clusters on umap embeddings\n",
    "sc.pl.draw_graph(adata_downsampled, color='leiden_1.0',\n",
    "                 legend_loc='on data', \n",
    "                 legend_fontsize = 'small',\n",
    "                #save='_n_neighbor20_UMAP_embedding_FA.png'\n",
    "                )\n",
    "\n",
    "#plot paga map\n",
    "sc.tl.paga(adata_downsampled, groups='leiden_1.0')\n",
    "sc.pl.paga(adata_downsampled, color='leiden_1.0', \n",
    "           edge_width_scale = 0.7, \n",
    "           node_size_scale=3,\n",
    "           #save='_n_neighbor20_UMAP_embedding_PAGA.png'\n",
    "          )"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "19ac2757-baab-41c4-9191-a66f62d144a8",
   "metadata": {},
   "source": [
    "#### Lines between course-grained paga clusters represents relative relatedness (i.e. how many genes are shared between clusters)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "5fe47c38-ce23-4a2f-80be-6bcd19b82218",
   "metadata": {},
   "source": [
    "## Filtering graph edges\n",
    "First, lets explore the graph a bit. So we plot the umap with the graph connections on top."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "3a37b8e4-5cf7-4a41-a4ab-8f44718496f0",
   "metadata": {},
   "outputs": [],
   "source": [
    "sc.pl.umap(adata_downsampled, edges=True, color = 'leiden_1.0', \n",
    "           legend_loc= 'on data', legend_fontsize= 'xx-small',\n",
    "           #save=\"_leiden_1.0_20n.neighbors_edges.png\"\n",
    "          )"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "a9f6571c-c1a2-4605-893e-523de11b7bd9",
   "metadata": {},
   "source": [
    "#### We have many edges in the graph between unrelated clusters, so lets try with fewer neighbors."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "dc0384fa-5693-45bd-b178-90ca02d4a19a",
   "metadata": {},
   "outputs": [],
   "source": [
    "sc.pp.neighbors(adata_downsampled, n_neighbors=5, n_pcs = 40) #ran 40pcs before\n",
    "sc.pl.umap(adata_downsampled, edges=True, color = 'leiden_1.0', \n",
    "           legend_loc= 'on data', legend_fontsize= 'xx-small',\n",
    "           #save=\"_leiden_1.0_5n.neighbors_edges.png\"\n",
    "          )"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "a8d8d252-bc9e-4859-9405-ec1f5c081931",
   "metadata": {},
   "source": [
    "## Rerun PAGA after lowering n_neighbors"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "7dda0440-3ad9-4e04-856e-f6d1f85fb256",
   "metadata": {},
   "outputs": [],
   "source": [
    "#use the umap to initialize the graph layout and use FA2 to spatialize\n",
    "#sc.tl.umap(adata_downsampled, min_dist=0.4, spread=3)\n",
    "#sc.tl.draw_graph(adata_downsampled, init_pos='X_umap')\n",
    "sc.pl.draw_graph(adata_downsampled, color='leiden_1.0', legend_loc='on data', \n",
    "                 legend_fontsize=  'xx-small', edges = True, #save = '_n_neighbor5_UMAP_embedding_FA.png'\n",
    "                )\n",
    "\n",
    "#then rerun paga\n",
    "#sc.tl.paga(adata_downsampled, groups='leiden_1.0')\n",
    "sc.pl.paga(adata_downsampled, color='leiden_1.0', edge_width_scale = 0.5, #node_size_scale= 2, save='_n_neighbor5_UMAP_embedding_PAGA.png'\n",
    "          )"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "a968c527-6e1d-4880-9a17-8f65f84e3769",
   "metadata": {},
   "source": [
    "## Embedding using PAGA-initialization\n",
    "We can now redraw the graph using the paga layout itself. The following is just as well possible for a UMAP or FA2 layout."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "5a4c4ec8-72b2-443c-bd1c-42734055af15",
   "metadata": {},
   "outputs": [],
   "source": [
    "sc.tl.draw_graph(adata_downsampled, init_pos='paga')"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "fb6f0154-fe00-45bb-b468-2c42f1fe2e64",
   "metadata": {},
   "source": [
    "Now we can see all marker genes also at single-cell resolution in the paga-oriented layout."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "210a948f-f325-4b4f-8b16-6339fcc0f6f9",
   "metadata": {},
   "outputs": [],
   "source": [
    "sc.pl.umap(adata_downsampled, color=['annot'], legend_loc='on data', legend_fontsize=  'xx-small', edges= True), #save='_paga_embed_n_neighbor5_umap.png')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "ee09826a-f0e0-4c3e-8ec7-179190bc87fc",
   "metadata": {},
   "outputs": [],
   "source": [
    "\n",
    "sc.pl.draw_graph(adata_downsampled, color=['leiden_1.0'], legend_loc='on data', legend_fontsize=  'xx-small', edges= True, save='_paga_embed_n_neighbor5_fa.png')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "07feeb05-d662-4ab5-9ca7-6276c3aa3a2e",
   "metadata": {},
   "outputs": [],
   "source": [
    "sc.pl.paga(adata_downsampled, color='leiden_1.0', edge_width_scale = 0.5, node_size_scale= 2, save='_n_neighbor5_paga_embed_paga.png')"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "cf73ac70-6560-4572-923f-8f8b770d2610",
   "metadata": {},
   "source": [
    "#### Compare UMAP and PAGA based on PAGA embedding"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "c5750a60-3d76-4f45-b920-9ab3c83ccba1",
   "metadata": {},
   "outputs": [],
   "source": [
    "sc.pl.paga_compare(\n",
    "    adata_downsampled, threshold=0.05, title='leiden_1.0',\n",
    "    size=10, edge_width_scale=0.5,\n",
    "    legend_fontsize=12, fontsize=12, frameon=True, edges=True, save='_paga_fa_compare_leiden_1.0.png')"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "da60aee3-daf4-42a6-a1db-85647fc7274f",
   "metadata": {
    "jp-MarkdownHeadingCollapsed": true
   },
   "source": [
    "#### testing 11/22/2025"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "9ad22dcf-cd18-4d4b-b10a-e7bf3614a0c5",
   "metadata": {},
   "outputs": [],
   "source": [
    "#DO NOT RUN: #nothing too convincing here...\n",
    "#probably not the best data to try this since there is only 2 biological replicates that were sequenced differently\n",
    "\n",
    "#Rename clusters with really clear markers that we ID's from our bulk analyses and from previous 'omics papers\n",
    "\n",
    "#for leiden_1.0 annots\n",
    "annot = pd.DataFrame(adata_downsampled.obs['leiden_1.0'].astype('string'))\n",
    "annot[annot['leiden_1.0'] == '9'] = '9_hyp_hcb' #hypothetical hematopoietic stem cells based on HHEX and other marker expression\n",
    "adata_downsampled.obs['annot']=annot['leiden_1.0'].astype('category')\n",
    "sc.pl.umap(adata_downsampled, color = 'annot',legend_loc = 'on data', legend_fontsize = 'xx-small', ncols =2)\n",
    "annot.value_counts()\n",
    "type(annot)\n",
    "\n",
    "\n",
    "#Pseudotime with Monocle and hopping back to the first tutorial:\n",
    "#https://nbisweden.github.io/workshop-scRNAseq/labs/scanpy/scanpy_07_trajectory.html)\n",
    "\n",
    "#We can reconstruct gene changes along PAGA paths for a given set of genes\n",
    "#Choose a root cell for diffusion pseudotime.\n",
    "\n",
    "#docs say to first run diffmap then do dtp\n",
    "#https://scanpy.readthedocs.io/en/stable/generated/scanpy.tl.dpt.html\n",
    "#https://scanpy.readthedocs.io/en/stable/generated/scanpy.tl.diffmap.html#scanpy.tl.diffmap\n",
    "\n",
    "\n",
    "#need to use gauss and not umap - In order to reproduce the original implementation of DPT, use method=='gauss'. Using the default method=='umap' only leads to minor quantitative differences, though.\n",
    "sc.pp.neighbors(adata_downsampled, n_neighbors = 5, #use latest settings\n",
    "                n_pcs = 40, method = 'gauss', use_rep = \"X_harmony\")\n",
    "adata_downsampled.uns['iroot'] = np.flatnonzero(adata_downsampled.obs['annot']  == '9_hyp_hcb')[0]\n",
    "\n",
    "\n",
    "#As previously, dpt() came with a default parameter of n_dcs=10 but diffmap() has a default parameter of n_comps=15, \n",
    "#you need to pass n_comps=10 in diffmap() in order to exactly reproduce previous dpt() results. \n",
    "sc.tl.diffmap(adata_downsampled, n_comps=10),\n",
    "                  #neighbors_key=None, random_state=0, copy=False)\n",
    "\n",
    "sc.tl.dpt(adata_downsampled)#, n_dcs=10, n_branchings=0, \n",
    "              #min_group_size=0.01, allow_kendall_tau_shift=True, \n",
    "              #neighbors_key=None, copy=False)\n",
    "\n",
    "\n",
    "sc.pl.draw_graph(adata_downsampled, color=['annot', 'dpt_pseudotime'], legend_loc='on data', legend_fontsize= 'x-small')\n",
    "\n",
    "#am able to get pseudotime to run but without canonical markers to clusters it's moot"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "84d80754-aa61-4d9f-a9dc-056c22f57b05",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "16f251b8-033c-4572-a0cd-0830420e0d2f",
   "metadata": {},
   "outputs": [],
   "source": [
    "sc.pl.correlation_matrix(adata_downsampled, groupby=\"leiden_1.0\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "602b166c-f5b1-4f71-b057-0c33dc27c164",
   "metadata": {},
   "outputs": [],
   "source": [
    "sc.pl.paga(adata_downsampled, layout='rt', root=[9])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "4f360df8-15f2-4776-920f-ad9ba1b273a1",
   "metadata": {},
   "outputs": [],
   "source": [
    "sc.pl.highest_expr_genes(adata_downsampled, n_top=30)\n",
    "sc.pl.highly_variable_genes(adata_downsampled)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "e633bc8d-e9f0-4581-b46b-f5e6c78577d9",
   "metadata": {},
   "outputs": [],
   "source": [
    "sc.pl.diffmap(adata_downsampled, color = 'leiden_1.0')"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "d5eacc70-ecc3-4144-ae57-0b5aa380adf4",
   "metadata": {},
   "source": [
    "## Further pruning PAGA edges \n",
    "Again from this tutorial -> https://scanpy.readthedocs.io/en/latest/tutorials/plotting/advanced.html\n",
    "\n",
    "Remove weak PAGA edges in the plot based on edge weight distribution.\n",
    "\n",
    "This is based on the assumption that most edge weights will be relatively weak and that we can therefore spot the few most interesting edges as the outliers of the distribution."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "736bdc56-4d7e-408c-bf2a-3ba628755b06",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Distribution of PAGA connectivities for determining the cutting threshold\n",
    "fig, axs = plt.subplots(1, 2, figsize=(6, 3))\n",
    "paga_conn = adata_downsampled.uns[\"paga\"][\"connectivities\"].toarray().ravel()\n",
    "a = axs[0].hist(paga_conn, bins=30)\n",
    "sns.violinplot(paga_conn, ax=axs[1], inner=None)\n",
    "sns.swarmplot(paga_conn, ax=axs[1], color=\"k\", size=1)\n",
    "thr = .5\n",
    "_ = axs[1].axhline(thr, c=\"r\")\n",
    "_ = axs[0].axvline(thr, c=\"r\")\n",
    "#fig.savefig(\"/Users/johnbriseno/Desktop/M_WBs_Jupyter/figures/M_WBs_paga_pruning_qc.png\", format=\"png\")\n",
    "fig.show()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "7d72300d-b519-4412-a9b8-b8d1cc572c7f",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Compare PAGA with and without prunning\n",
    "fig, axs = plt.subplots(1, 2, figsize=(6, 3))\n",
    "sc.pl.paga(adata_downsampled, ax=axs[0], title=\"PAGA\", show=False)\n",
    "sc.pl.paga(adata_downsampled, ax=axs[1], title=\"PAGA - prunned\", threshold=thr, save = 'paga_pruning.png')"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "e0fb7323-b807-434d-a048-67010aa77871",
   "metadata": {},
   "source": [
    "## PAGA super-imposed on UMAP and FA\n",
    "The layout used in PAGA is optimised to correspond to the PAGA connectivties (edge weighs). However, sometimes we would wish to have a different layout. For this we can use the pos argument."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "9df3a61f-b4cc-4ede-9178-14e324fa70eb",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Compare UMAP and PAGA layouts\n",
    "fig, axs = plt.subplots(1, 2, figsize=(6, 3))\n",
    "sc.pl.umap(adata_downsampled, color=\"leiden_1.0\", ax=axs[0], show=False, title=\"UMAP\", legend_loc=\"on data\")\n",
    "sc.pl.paga(adata_downsampled, ax=axs[1], title=\"PAGA\")"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "78e6aca4-3c74-43f0-b13f-48a09ae60b34",
   "metadata": {},
   "source": [
    "#### PAGA layout corresponding to UMAP\n",
    "Set PAGA dot centers to the mean of the UMAP embedding values of cells from the corresponding groups."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "e0b7934b-efa0-4962-8f49-ffc8a5464668",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Define PAGA positions based on the UMAP layout -\n",
    "#for each cluster we use the mean of the UMAP positions from the cells in that cluster\n",
    "pos = pd.DataFrame(adata_downsampled.obsm[\"X_umap\"], index=adata_downsampled.obs_names)\n",
    "pos[\"group\"] = adata_downsampled.obs[adata_downsampled.uns[\"paga\"][\"groups\"]]\n",
    "pos = pos.groupby(\"group\", observed=True).mean()\n",
    "\n",
    "# Plot UMAP in the background\n",
    "ax = sc.pl.umap(adata_downsampled, show=False)\n",
    "# Plot PAGA ontop of the UMAP\n",
    "with rc_context({\"figure.figsize\": (5, 5)}):\n",
    "    sc.pl.paga(\n",
    "        adata_downsampled,\n",
    "        color=\"leiden_1.0\",\n",
    "        threshold=thr,\n",
    "        node_size_scale=2,\n",
    "        edge_width_scale=.5,\n",
    "        pos=pos.values,\n",
    "        random_state=0,\n",
    "        ax=ax,\n",
    "        frameon=False,\n",
    "        #save='paga_on_umap.svg'\n",
    ")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "6350b443-6588-475d-b00f-1c1955ce9c68",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Define PAGA positions based on the FA layout -\n",
    "#for each cluster we use the mean of the fa positions from the cells in that cluster\n",
    "pos = pd.DataFrame(adata_downsampled.obsm[\"X_draw_graph_fa\"], index=adata_downsampled.obs_names)\n",
    "pos[\"group\"] = adata_downsampled.obs[adata_downsampled.uns[\"paga\"][\"groups\"]]\n",
    "pos = pos.groupby(\"group\", observed=True).mean()\n",
    "\n",
    "# Plot fa in the background\n",
    "ax = sc.pl.draw_graph(adata_downsampled, show=False)\n",
    "# Plot PAGA ontop of the UMAP\n",
    "sc.pl.paga(\n",
    "    adata_downsampled,\n",
    "    color=\"leiden_1.0\",\n",
    "    threshold=thr,\n",
    "    node_size_scale=1,\n",
    "    edge_width_scale=0.5,\n",
    "    pos=pos.values,\n",
    "    random_state=0,\n",
    "    ax=ax,\n",
    "    frameon=False,\n",
    "    save='paga_on_fa.svg'\n",
    "\n",
    ")"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "caf89e7c-a5f6-4a59-af9f-fcacd50073b1",
   "metadata": {},
   "source": [
    "#### Showing different spreads and min_dists for UMAP\n",
    "from: https://scanpy.readthedocs.io/en/stable/tutorials/plotting/advanced.html"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "94e426a6-8f3a-4540-9510-c5bbbb188319",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Copy adata not to modify UMAP in the original adata object\n",
    "adata_temp = adata_downsampled.copy()\n",
    "# Loop through different umap parameters, recomputting and replotting UMAP for each of them\n",
    "MIN_DISTS = [0.1, 1, 2]\n",
    "SPREADS = [0.5, 1, 5]\n",
    "\n",
    "# Create grid of plots, with a little extra room for the legends\n",
    "fig, axes = plt.subplots(\n",
    "    len(MIN_DISTS), len(SPREADS), figsize=(len(SPREADS) * 3 + 2, len(MIN_DISTS) * 3)\n",
    ")\n",
    "for (i, min_dist), (j, spread) in product(enumerate(MIN_DISTS), enumerate(SPREADS)):\n",
    "    ax = axes[i][j]\n",
    "    param_str = \" \".join([\"min_dist =\", str(min_dist), \"and spread =\", str(spread)])\n",
    "    # Recompute UMAP with new parameters\n",
    "    sc.tl.umap(adata_temp, min_dist=min_dist, spread=spread)\n",
    "    # Create plot, placing it in grid\n",
    "    sc.pl.umap(\n",
    "        adata_temp,\n",
    "        color=[\"leiden_1.0\"],\n",
    "        title=param_str,\n",
    "        s=40,\n",
    "        ax=ax,\n",
    "        show=False,\n",
    "    )\n",
    "plt.tight_layout()\n",
    "plt.show()\n",
    "\n",
    "fig.savefig(\"_paga_embed_umap_diff_parameters.png\", format=\"png\")\n",
    "#fig.show()\n",
    "plt.close()\n",
    "del adata_temp\n",
    "\n",
    "#annot\n",
    "#seurat_clusters"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "3e40b5c7-c4b5-4d10-8067-8a42a0376843",
   "metadata": {
    "jp-MarkdownHeadingCollapsed": true
   },
   "source": [
    "#### Producing nicer UMAPs/FA with PAGA embedding"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "a699d28c-6c3f-48ce-ae95-958fd6bf2de2",
   "metadata": {},
   "outputs": [],
   "source": [
    "#sc.tl.draw_graph(adata_downsampled, init_pos='umap')\n",
    "\n",
    "#with rc_context({\"figure.figsize\": (5, 5)}):\n",
    "sc.pl.umap(\n",
    "        adata_downsampled,\n",
    "        color=\"leiden_1.0\",\n",
    "        add_outline=True,\n",
    "        legend_loc=\"on data\",\n",
    "        legend_fontsize=10,\n",
    "        legend_fontoutline=2,\n",
    "        frameon=False,\n",
    "        title='',\n",
    "        edges=True,\n",
    "        save='_n_neighbor5_paga_embed_umap.svg'\n",
    "    )"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "06b6eca2-0c1d-4a4b-8c09-710681c0040e",
   "metadata": {},
   "outputs": [],
   "source": [
    "sc.pl.umap(\n",
    "        adata_downsampled,\n",
    "        color=\"leiden_1.0\",\n",
    "        add_outline=True,\n",
    "        legend_loc=\"on data\",\n",
    "        legend_fontsize=10,\n",
    "        legend_fontoutline=2,\n",
    "        frameon=False,\n",
    "        title='',\n",
    "        edges=True,\n",
    "        save='_n_neighbor5_paga_embed_umap.svg'\n",
    "    )"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "18f995f0-1176-4e68-9665-49f224f89ee3",
   "metadata": {},
   "outputs": [],
   "source": [
    "#sc.tl.draw_graph(adata_downsampled, init_pos='X_draw_graph_fa')\n",
    "\n",
    "#with rc_context({\"figure.figsize\": (5, 5)}):\n",
    "sc.pl.draw_graph(\n",
    "        adata_downsampled,\n",
    "        color=\"leiden_1.0\",\n",
    "        add_outline=True,\n",
    "        legend_loc=\"on data\",\n",
    "        legend_fontsize=10,\n",
    "        legend_fontoutline=2,\n",
    "        frameon=False,\n",
    "        edges=True,\n",
    "        title='',\n",
    "        save='_n_neighbor5_paga_embed_fa.svg',\n",
    "    )"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "5cdb1631-5567-4581-80b4-ec9c6cc06a08",
   "metadata": {},
   "outputs": [],
   "source": [
    "axes_dict = vp.get_axes()\n",
    "print(axes_dict)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "b0b9c418-3ffe-4c20-946a-221411bf214d",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "d56e2595-c625-4be3-b893-be4b0a76faf6",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "8ce64d62-69f1-465d-a13d-f01f550ab5f8",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "d6303bc4-1d25-4d01-a3c8-0ea35cc113fe",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "bcfd1673-9768-45c1-b6ed-5ce91ca5208a",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "markdown",
   "id": "b58bb47b-5729-42a6-9db3-783f2696e959",
   "metadata": {},
   "source": [
    "#### assigning clusters based on bulk markers as evidence - we did not run in our analyses"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "3552e3c1-d01a-4612-a590-9bd06b1d3715",
   "metadata": {},
   "outputs": [],
   "source": [
    "#Assigning names to clusters with really clear markers that we ID's from our bulk analyses and from previous 'omics papers\n",
    "#for leiden_1.0 annots\n",
    "annot = pd.DataFrame(adata.obs['leiden_1.0'].astype('string'))\n",
    "annot[annot['leiden_1.0'] == '8'] = '8_hyp_hsc' #hypothetical hematopoietic stem cells based on HHEX and other marker expression\n",
    "adata.obs['annot']=annot['leiden_1.0'].astype('category')\n",
    "sc.pl.umap(adata, color = 'annot',legend_loc = 'on data', legend_fontsize = 'xx-small', ncols =2)\n",
    "annot.value_counts()\n",
    "type(annot)\n",
    "\n",
    "#for seurat_clusters annots\n",
    "annot = pd.DataFrame(adata.obs['seurat_clusters'].astype('string'))\n",
    "annot[annot['seurat_clusters'] == '6'] = '6_hyp_hsc' #hypothetical hematopoietic stem cells based on HHEX and other marker expression\n",
    "adata.obs['annot']=annot['seurat_clusters'].astype('category')\n",
    "sc.pl.umap(adata, color = 'annot',legend_loc = 'on data', legend_fontsize = 'xx-small', ncols =2)\n",
    "annot.value_counts()\n",
    "type(annot)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "eab9ae47-dd22-4caf-88bb-8264bbb88703",
   "metadata": {},
   "source": [
    "We can run monocle for pseudotime, but only once we start seeing good trends in data. Our data is OK but not seeing too much relevancy. "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "39920832-7700-4c33-8a2f-3a2ac33d3f99",
   "metadata": {},
   "outputs": [],
   "source": [
    "#DO NOT RUN: #nothing too convincing here...\n",
    "#probably not the best data to try this since there is only 2 biological replicates that were sequenced differently\n",
    "\n",
    "#Pseudotime with Monocle and hopping back to the first tutorial:\n",
    "#https://nbisweden.github.io/workshop-scRNAseq/labs/scanpy/scanpy_07_trajectory.html)\n",
    "\n",
    "#We can reconstruct gene changes along PAGA paths for a given set of genes\n",
    "#Choose a root cell for diffusion pseudotime.\n",
    "\n",
    "#docs say to first run diffmap then do dtp\n",
    "#https://scanpy.readthedocs.io/en/stable/generated/scanpy.tl.dpt.html\n",
    "#https://scanpy.readthedocs.io/en/stable/generated/scanpy.tl.diffmap.html#scanpy.tl.diffmap\n",
    "\n",
    "\n",
    "#need to use gauss and not umap - In order to reproduce the original implementation of DPT, use method=='gauss'. Using the default method=='umap' only leads to minor quantitative differences, though.\n",
    "sc.pp.neighbors(adata, n_neighbors = 5, #use latest settings\n",
    "                n_pcs = 41, method = 'gauss', use_rep = \"X_harmony\")\n",
    "adata.uns['iroot'] = np.flatnonzero(adata.obs['annot']  == '8_hyp_hsc')[0]\n",
    "\n",
    "\n",
    "#As previously, dpt() came with a default parameter of n_dcs=10 but diffmap() has a default parameter of n_comps=15, \n",
    "#you need to pass n_comps=10 in diffmap() in order to exactly reproduce previous dpt() results. \n",
    "sc.tl.diffmap(adata, n_comps=10),\n",
    "                  #neighbors_key=None, random_state=0, copy=False)\n",
    "\n",
    "sc.tl.dpt(adata)#, n_dcs=10, n_branchings=0, \n",
    "              #min_group_size=0.01, allow_kendall_tau_shift=True, \n",
    "              #neighbors_key=None, copy=False)\n",
    "\n",
    "\n",
    "sc.pl.draw_graph(adata, color=['annot', 'dpt_pseudotime'], legend_loc='on data', legend_fontsize= 'x-small')"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "1b0d8d48-8730-4617-b3bf-215ce000fcc4",
   "metadata": {},
   "source": [
    "## 3D UMAP changing n_neighbor values\n",
    "based on https://github.com/MNoichl/UMAP-examples-mammoth-/tree/master"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "a5bae9b3-b848-4fb8-aa0c-72f3facefc39",
   "metadata": {},
   "outputs": [],
   "source": [
    "import pandas as pd\n",
    "import numpy as np\n",
    "import umap\n",
    "%matplotlib inline\n",
    "\n",
    "import seaborn as sns\n",
    "import matplotlib.pyplot as plt\n",
    "import matplotlib as mpl\n",
    "#For Tables:\n",
    "from IPython.display import display\n",
    "from IPython.display import Latex\n",
    "pd.set_option('display.max_columns', 500)\n",
    "from tqdm import tnrange, tqdm_notebook\n",
    "%load_ext rpy2.ipython\n",
    "\n",
    "     \n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "62082740-810f-4ba7-b0ed-56a863a13f1c",
   "metadata": {},
   "outputs": [],
   "source": [
    "#3D coordinates\n",
    "sc.tl.umap(adata_downsampled, n_components=3)\n",
    "sn_umap = adata_downsampled.obsm['X_umap']\n",
    "adata.obsm['X_umap']"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "d7e226ee-18de-4ae0-8724-27d94ce08dcb",
   "metadata": {},
   "outputs": [],
   "source": [
    "#plot the umap\n",
    "from mpl_toolkits.mplot3d import Axes3D\n",
    "\n",
    "fig = plt.figure(figsize=(90,72))\n",
    "\n",
    "ax = fig.add_subplot(111, projection='3d')\n",
    "# ax.set_aspect('equal')\n",
    "\n",
    "ax.set_axis_off()\n",
    "ax.scatter(sn_umap[:,0], sn_umap[:,1], sn_umap[:,2], s=20,c='black')\n",
    "ax.view_init(0, -170)\n",
    "\n",
    "plt.show()\n",
    "\n",
    "#works"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "370cb9e0-f8e4-42da-8149-4a1fc0bd6d58",
   "metadata": {},
   "outputs": [],
   "source": [
    "reducer = umap.UMAP(random_state=42,\n",
    "                    init='random',\n",
    "                    n_components=3,\n",
    "                    n_neighbors=300,\n",
    "                    min_dist=0.1,\n",
    "                    spread=2,\n",
    "                    metric='euclidean',\n",
    "                    verbose=True)\n",
    "\n",
    "reducer = reducer.fit(sn_umap)\n",
    "#works"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "76c0d81b-2cb6-4932-8a0a-14b5de4aa7d4",
   "metadata": {},
   "outputs": [],
   "source": [
    "plt.figure(figsize=(40,40),facecolor='w')\n",
    "\n",
    "\n",
    "#fig, ax = plt.subplots(facecolor='w')\n",
    "plt.axis('off')\n",
    "plt.scatter(reducer.embedding_[:, 0], reducer.embedding_[:, 1], s=2.5,c='black')\n",
    "\n",
    "#ax = plt.Axes(fig, [0., 0., 1., 1.])\n",
    "#ax.set_axis_off()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "62bf8d2d-c8a1-417e-960c-af30f6c4f81b",
   "metadata": {},
   "outputs": [],
   "source": [
    "fig = plt.figure(figsize=(40,40),facecolor='w')\n",
    "ax = plt.subplots(facecolor='w')\n",
    "plt.axis('off')\n",
    "\n",
    "#2D\n",
    "#plt.scatter(reducer.embedding_[:, 0], reducer.embedding_[:, 1], s=2.5, c='black')\n",
    "\n",
    "#3D\n",
    "plt.scatter(reducer.embedding_[:, 0], reducer.embedding_[:, 1], reducer.embedding_[:, 2], #s=2.5, \n",
    "            c='black')\n",
    "\n",
    "\n",
    "#ax = plt.Axes(fig, [0., 0., 1., 1.])\n",
    "#ax.set_axis_off()\n",
    "##works"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "547fa0e3-fc14-4545-be44-8a73a1e67b02",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Perform agglomerative clustering on a subset, then scale up with a classifier. Courtesy sb. on StackOverflow. \n",
    "from sklearn.cluster import AgglomerativeClustering\n",
    "clustering = AgglomerativeClustering(n_clusters=15).fit(sn_umap) #change to however many clusters you're working with \n",
    "from sklearn.model_selection import train_test_split\n",
    "from sklearn.neighbors import KNeighborsClassifier\n",
    "\n",
    "X_train, X_test, y_train, y_test = \\\n",
    "    train_test_split(sn_umap, sn_umap, \n",
    "                     test_size=6000, random_state=22) \n",
    "AC = AgglomerativeClustering(n_clusters=15, linkage='ward')\n",
    "AC.fit(X_train)\n",
    "labels = AC.labels_\n",
    "\n",
    "KN = KNeighborsClassifier(n_neighbors=10)\n",
    "KN.fit(X_train,labels)\n",
    "labels2 = KN.predict(sn_umap)\n",
    "col_len = len(set(labels2))-1\n",
    "\n",
    "#works"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "e326c5c3-07ae-49e7-b355-5901a3321457",
   "metadata": {},
   "outputs": [],
   "source": [
    "%%R -i col_len -o color_scale\n",
    "\n",
    "mycolors <- c('#1f77b4',\n",
    " '#ff7f0e',\n",
    " '#279e68',\n",
    " '#d62728',\n",
    " '#aa40fc',\n",
    " '#8c564b',\n",
    " '#e377c2',\n",
    " '#b5bd61',\n",
    " '#17becf',\n",
    " '#aec7e8',\n",
    " '#ffbb78',\n",
    " '#98df8a',\n",
    " '#ff9896',\n",
    " '#c5b0d5',\n",
    " '#c49c94')\n",
    "pal <- colorRampPalette(sample(mycolors))\n",
    "color_scale <- sample(pal(col_len))\n",
    "color_scale <- c(color_scale)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "16512a97-d880-485c-81f0-151bec3f03b1",
   "metadata": {},
   "outputs": [],
   "source": [
    "cmap = mpl.colors.ListedColormap(list(color_scale))\n",
    "sns.palplot (color_scale)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "fbc77c3f-3c7d-4e3b-affc-5f28b1aff07a",
   "metadata": {},
   "outputs": [],
   "source": [
    "cmap = mpl.colors.ListedColormap(adata_downsampled.uns['leiden_1.0_colors'])\n",
    "cmap\n",
    "#sns.palplot(color_scale)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "2b39814f-96c1-4357-98c4-fa549c8bebe4",
   "metadata": {},
   "outputs": [],
   "source": [
    "from mpl_toolkits.mplot3d import Axes3D\n",
    "\n",
    "fig = plt.figure(figsize=(48,35))\n",
    "\n",
    "ax = fig.add_subplot(111, projection='3d')\n",
    "# ax.set_aspect('equal')\n",
    "\n",
    "ax.set_axis_off()\n",
    "ax.scatter(sn_umap[:,0], sn_umap[:,1], sn_umap[:,2], s=5, c=labels2, cmap=cmap)\n",
    "ax.view_init(10, -170)\n",
    "\n",
    "plt.show()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "d1b85e43-a788-4520-af02-b5765c0e1a45",
   "metadata": {},
   "outputs": [],
   "source": [
    "cmap"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "efc611de-0c4d-44e8-8a9d-849f07a4d49b",
   "metadata": {},
   "outputs": [],
   "source": [
    "from mpl_toolkits.mplot3d import Axes3D\n",
    "\n",
    "fig = plt.figure(figsize=(48,30))\n",
    "\n",
    "ax = fig.add_subplot(111, projection='3d')\n",
    "# ax.set_aspect('equal')\n",
    "\n",
    "ax.set_axis_off()\n",
    "ax.scatter(sn_umap[:,0], sn_umap[:,1], sn_umap[:,2], s=5, c=labels2, cmap=cmap)\n",
    "ax.view_init(90, 0)\n",
    "\n",
    "plt.show()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "02f6346b-2e48-4d82-bad3-94153344237d",
   "metadata": {},
   "outputs": [],
   "source": [
    "plt.figure(figsize=(40,40),facecolor='w')\n",
    "\n",
    "plt.axis('off')\n",
    "plt.scatter(reducer.embedding_[:, 0], reducer.embedding_[:, 1], s=5,c=labels2,cmap=cmap)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "b9bb6756-e781-4491-a199-800c0340f95b",
   "metadata": {},
   "outputs": [],
   "source": [
    "#Changing n_neighbours\n",
    "#Loop over several values of n_neighbours and always use the previous embedding as the initialization of the new one instead of a spectral embedding. Slightly reduce the learning rate, so that this works nicely.\n",
    "\n",
    "neighbours = list(range(5,120,5))\n",
    "neighbours.reverse()\n",
    "first = True\n",
    "coordinates = []\n",
    "for n in tqdm_notebook(neighbours):\n",
    "    #min_dist=0.1,\n",
    "\n",
    "    if first == True:\n",
    "        first = False\n",
    "        reducer = umap.UMAP(n_components=2,n_neighbors=n,\n",
    "                        metric='euclidean')\n",
    "        reducer.fit(sn_umap)\n",
    "    else:\n",
    "        reducer = umap.UMAP(n_components=2,n_neighbors=n,\n",
    "                        metric='euclidean',learning_rate=0.9,init=embedding)\n",
    "        reducer.fit(sn_umap)\n",
    "    embedding = reducer.embedding_\n",
    "\n",
    "    plt.figure(figsize=(6,6),facecolor='w')\n",
    "    plt.axis('off')\n",
    "    plt.scatter(reducer.embedding_[:, 0], reducer.embedding_[:, 1], s=2.5,c=labels2,cmap=cmap)#color='black' doesnt need labesl2\n",
    "    emb = pd.DataFrame(embedding)\n",
    "    emb.columns = ['x','y']\n",
    "    emb[\"n\"] = n\n",
    "    coordinates.append(emb)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "3e06b895-cb4b-4852-b244-4f3fb41dbedf",
   "metadata": {},
   "outputs": [],
   "source": [
    "coordinates_full_df = pd.concat(coordinates, ignore_index=True)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "dcbe05cb-8099-4e23-b5de-55c3aff54be5",
   "metadata": {},
   "outputs": [],
   "source": [
    "%%R -i coordinates_full_df\n",
    "#--width 4900 --height 3200 -r 140 --bg #ffffff\n",
    "library(hrbrthemes)\n",
    "library(ggplot2)\n",
    "library(fields)\n",
    "library(akima)\n",
    "library(directlabels)\n",
    "library(zoo)\n",
    "library(gridExtra)\n",
    "library(plotly)\n",
    "\n",
    "library(ggalt)\n",
    "library(mgcv)\n",
    "library(scales)\n",
    "library(gganimate)\n",
    "library(gifski)\n",
    "library(png)\n",
    "\n",
    "timespan <- length(unique(coordinates_full_df$n))\n",
    "\n",
    "p <- ggplot(data = coordinates_full_df,aes(x=x,y=y))+#,group=id,color=as.factor(journal)))+#as.factor(id)))+#))+\n",
    "# geom_point(data=subset(coordinates_full_df, cluster == -1), aes(x=x, y=y),pch=16,cex=1.5,alpha=0.2, color=myGray)+                      \n",
    "# geom_point(data=subset(coordinates_full_df, cluster != -1), aes(x=x, y=y,color=as.factor(cluster)),pch=16,cex=1.5,alpha=0.6)+                      \n",
    "\n",
    "geom_point(size = 0.9, stroke = 0,alpha=1)+\n",
    "\n",
    "theme_ipsum_rc()+\n",
    "theme(legend.position = \"none\")+\n",
    "labs(x=\"\", y=\"\",\n",
    "       title=\"UMAP of a Woolly Mammoth\",\n",
    "#        subtitle=\"...based on the code by McInnes, Healy (2018)\",\n",
    "       caption=\"by Maximilian Noichl, UMAP by McInnes, Healy (2018), Mammooth by Smithsonian 3D\")+\n",
    "theme(panel.grid.major = element_line(colour = \"grey\", linetype=\"dotted\", size=0.55),panel.grid.minor = element_blank())+\n",
    " theme(plot.background = element_rect(fill = \"#fbf8f1\"))+\n",
    "\n",
    "  transition_states(n) +\n",
    "  ease_aes('sine-in-out', interval = 0.01)+#,enter_length = 0.002,exit_length = 0.002)+,nframes = 1000\n",
    "enter_appear(early = FALSE)+\n",
    "#  enter_fade()+\n",
    "#    exit_fade()+\n",
    "exit_disappear(early = TRUE)+\n",
    "labs(subtitle = \"Embedding for n_neighbours = {closest_state}\")\n",
    "\n",
    "# p <- p + \n",
    "#   transition_time(year,transition_length = 4,state_length = 0)+\n",
    "#   labs(title = \"Year: {frame_time}\")\n",
    "p <-animate(p, \n",
    "        duration = timespan,#s/100, # = 365 days/yr x 3 years x 0.25 sec/day = 274 seconds\n",
    "        fps  =  28,width=1000,height=1000)#,nframes = [...or pick it here])\n",
    "\n",
    "\n",
    "\n",
    "print(p)\n",
    "anim_save('MWB_anim_nearest_neighbours.gif', animation = p)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "6b7bdaf6-82d3-477f-91a9-32f60face1f4",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "3cb6a99a-cc48-4428-b6aa-1376faf99c9b",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "76cc95c6-d8df-4091-8f1e-e0a2e1ab03ad",
   "metadata": {},
   "outputs": [],
   "source": [
    "#Do the same for min distances\n",
    "\n",
    "min_dists = [0.001,0.005,0.01,0.05,0.1,0.2,0.3,0.4,0.5,0.6,0.7,0.8]\n",
    "min_dists.reverse()\n",
    "min_dists\n",
    "first = True\n",
    "coordinates = []\n",
    "for d in tqdm_notebook(min_dists):\n",
    "    #min_dist=0.1,\n",
    "\n",
    "    if first == True:\n",
    "        first = False\n",
    "        reducer = umap.UMAP(n_components=2,n_neighbors=100,min_dist=d,\n",
    "                        metric='euclidean')\n",
    "        reducer.fit(sn_umap)\n",
    "    else:\n",
    "        reducer = umap.UMAP(n_components=2,n_neighbors=100,min_dist=d,\n",
    "                        metric='euclidean',learning_rate=0.9,init=embedding)#,learning_rate=0.9\n",
    "        reducer.fit(sn_umap)\n",
    "    embedding = reducer.embedding_\n",
    "\n",
    "    plt.figure(figsize=(6,6),facecolor='w')\n",
    "    plt.axis('off')\n",
    "    #plt.scatter(reducer.embedding_[:, 0], reducer.embedding_[:, 1], s=2.5,color='black')\n",
    "    plt.scatter(reducer.embedding_[:, 0], reducer.embedding_[:, 1], s=2.5,c=labels2,cmap=cmap)\n",
    "\n",
    "    emb = pd.DataFrame(embedding)\n",
    "    emb.columns = ['x','y']\n",
    "    emb[\"d\"] = d\n",
    "    coordinates.append(emb)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "1a0453eb-4e9d-4db7-a4ab-0d240534c7a1",
   "metadata": {},
   "outputs": [],
   "source": [
    "coordinates_full_df = pd.concat(coordinates, ignore_index=True)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "7962006e-c147-4b0d-ac4f-cbf0ff053ca7",
   "metadata": {},
   "outputs": [],
   "source": [
    "%%R -i coordinates_full_df\n",
    "#--width 4900 --height 3200 -r 140 --bg #ffffff\n",
    "library(hrbrthemes)\n",
    "library(ggplot2)\n",
    "library(fields)\n",
    "library(akima)\n",
    "library(directlabels)\n",
    "library(zoo)\n",
    "library(gridExtra)\n",
    "library(plotly)\n",
    "\n",
    "library(ggalt)\n",
    "library(mgcv)\n",
    "library(scales)\n",
    "library(gganimate)\n",
    "library(gifski)\n",
    "library(png)\n",
    "\n",
    "timespan <- length(unique(coordinates_full_df$d))\n",
    "\n",
    "p <- ggplot(data = coordinates_full_df,aes(x=x,y=y))+#,group=id,color=as.factor(journal)))+#as.factor(id)))+#))+\n",
    "# geom_point(data=subset(coordinates_full_df, cluster == -1), aes(x=x, y=y),pch=16,cex=1.5,alpha=0.2, color=myGray)+                      \n",
    "# geom_point(data=subset(coordinates_full_df, cluster != -1), aes(x=x, y=y,color=as.factor(cluster)),pch=16,cex=1.5,alpha=0.6)+                      \n",
    "\n",
    "geom_point(size = 0.9, stroke = 0,alpha=1)+\n",
    "\n",
    "theme_ipsum_rc()+\n",
    "theme(legend.position = \"none\")+\n",
    "labs(x=\"\", y=\"\",\n",
    "       title=\"UMAP of a Woolly Mammoth\",\n",
    "        subtitle=\"...based on the code by McInnes, Healy (2018)\",\n",
    "       caption=\"by Maximilian Noichl, UMAP by McInnes, Healy (2018), Mammooth by Smithsonian 3D\")+\n",
    "#theme(panel.grid.major = element_line(colour = \"grey\", linetype=\"dotted\", size=0.55),panel.grid.minor = element_blank())+\n",
    " theme(plot.background = element_rect(fill = \"#fbf8f1\"))+\n",
    "\n",
    "  transition_states(d) +\n",
    "  ease_aes('sine-in-out', interval = 0.01)+#,enter_length = 0.002,exit_length = 0.002)+,nframes = 1000\n",
    "enter_appear(early = FALSE)+\n",
    "  enter_fade()+\n",
    "    exit_fade()+\n",
    "exit_disappear(early = TRUE)+\n",
    "labs(subtitle = \"Embedding for min_dist = {closest_state}\")\n",
    "\n",
    "# p <- p + \n",
    "#   transition_time(year,transition_length = 4,state_length = 0)+\n",
    "#   labs(title = \"Year: {frame_time}\")\n",
    "p <- animate(p, \n",
    "        duration = timespan,#s/100, # = 365 days/yr x 3 years x 0.25 sec/day = 274 seconds\n",
    "        fps  =  28,width=1000,height=1000)#,nframes = [...or pick it here])\n",
    "print(p)\n",
    "anim_save('MWBs_dist_param.gif', animation = p)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "736b4a6b-3717-41b0-9b6d-40cbc268b4be",
   "metadata": {},
   "outputs": [],
   "source": [
    "#Do the same for random states\n",
    "\n",
    "reducer = umap.UMAP(#random_state=42,\n",
    "                     init='random',\n",
    "                    n_components=3,\n",
    "                    n_neighbors=300,\n",
    "                    min_dist=0.1,\n",
    "#                     spread=2,\n",
    "                    metric='euclidean',\n",
    "                    verbose=True)\n",
    "\n",
    "reducer = reducer.fit(sn_umap)\n",
    "#works\n",
    "\n",
    "first = True\n",
    "coordinates = []\n",
    "for d in tqdm_notebook(min_dists):\n",
    "    #min_dist=0.1,\n",
    "\n",
    "    if first == True:\n",
    "        first = False\n",
    "        reducer = umap.UMAP(n_components=2,n_neighbors=100,min_dist=.1, #keep umap constant\n",
    "                        metric='euclidean')\n",
    "        reducer.fit(sn_umap)\n",
    "    else:\n",
    "        reducer = umap.UMAP(n_components=2,n_neighbors=100,min_dist=.1,\n",
    "                        metric='euclidean',learning_rate=0.9,init=embedding)#,learning_rate=0.9\n",
    "        reducer.fit(sn_umap)\n",
    "    embedding = reducer.embedding_\n",
    "\n",
    "    plt.figure(figsize=(6,6),facecolor='w')\n",
    "    plt.axis('off')\n",
    "    #plt.scatter(reducer.embedding_[:, 0], reducer.embedding_[:, 1], s=2.5,color='black')\n",
    "    plt.scatter(reducer.embedding_[:, 0], reducer.embedding_[:, 1], s=2.5,c=labels2,cmap=cmap)\n",
    "\n",
    "    emb = pd.DataFrame(embedding)\n",
    "    emb.columns = ['x','y']\n",
    "    emb[\"d\"] = d\n",
    "    coordinates.append(emb)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "8e7488b0-36bf-4bc1-9f1d-773728d068d6",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "bfd6d8bb-78eb-4da3-8d64-52d1deb10619",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "a08f60ab-f985-462e-ad73-352677274cd4",
   "metadata": {},
   "outputs": [],
   "source": [
    "#Another 3D UMAP tutorial from SamBiomics on utube \n",
    "#https://www.youtube.com/watch?v=XzAryDJTi1M\n",
    "\n",
    "#For 3D a\n",
    "import scanpy as sc\n",
    "import pandas as pd\n",
    "import numpy as np\n",
    "import matplotlib.pyplot as plt\n",
    "\n",
    "#For more than 20 colors \n",
    "#!pip install distinctipy \n",
    "#https://distinctipy.readthedocs.io/en/latest/usage.html#Import-distinctipy\n",
    "import distinctipy\n",
    "from distinctipy import colorblind\n",
    "from distinctipy import examples\n",
    "\n",
    "import matplotlib.pyplot as plt"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "b2034854-3d7b-45aa-8692-6574d096da84",
   "metadata": {},
   "outputs": [],
   "source": [
    "type(adata_downsampled.uns['leiden_1.0_colors'])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "0b523ba4-d875-4722-a92b-b121edf73dca",
   "metadata": {},
   "outputs": [],
   "source": [
    "type(plt.cm.tab20(range(0,15)))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "5aeb0a51-7835-426b-a9c3-77002784dca7",
   "metadata": {},
   "outputs": [],
   "source": [
    "color = dict(zip(range(0,10), adata_downsampled.uns['leiden_1.0_colors']))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "8cec91de-8298-4240-a5f1-559ec6018c2d",
   "metadata": {},
   "outputs": [],
   "source": [
    "plt.cm.tab20(range(0,10))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "7449504f-9e5c-46d7-91a9-b5d3a5bb996b",
   "metadata": {},
   "outputs": [],
   "source": [
    "sc.tl.umap(adata_downsampled, n_components=3)\n",
    "umap = adata_downsampled.obsm['X_umap']\n",
    "umap"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "781f82c0-d689-4db2-baef-f01ce19e9ba5",
   "metadata": {},
   "outputs": [],
   "source": [
    "fig = plt.figure(figsize = (10,10))\n",
    "ax = fig.add_subplot(projection = '3d')\n",
    "\n",
    "ax.scatter(umap[:,0], umap[:,1], umap[:,2])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "50d4b0d3-c2ca-42a5-bd75-86e490804567",
   "metadata": {},
   "outputs": [],
   "source": [
    "fig = plt.figure(figsize = (10,10))\n",
    "ax = fig.add_subplot(projection = '3d')\n",
    "\n",
    "ax.scatter(umap[:,0],umap[:,1],umap[:,2], \n",
    "           c = adata_downsampled.obs['leiden_1.0'].astype('int').map(color))\n",
    "\n",
    "plt.show()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "6a6805af-1ecb-4da6-890b-702d6c8371b3",
   "metadata": {},
   "outputs": [],
   "source": [
    "!mkdir figs\n",
    "fig = plt.figure(figsize = (10,10))\n",
    "ax = fig.add_subplot(projection = '3d')\n",
    "\n",
    "ax.scatter(umap[:,0], umap[:,1], umap[:,2], \n",
    "           c = adata_downsampled.obs['leiden_1.0'].astype('int').map(color))\n",
    "\n",
    "x_center = (umap[:,0].max() + umap[:,0].min())/2\n",
    "y_center = (umap[:,1].max() + umap[:,1].min())/2\n",
    "z_center = (umap[:,2].max() + umap[:,2].min())/2\n",
    "\n",
    "ax.plot([x_center,x_center], [y_center, y_center], [umap[:,2].min() - 2, umap[:,2].max() + 2], c = 'k', lw = 5)\n",
    "ax.plot([x_center,x_center], [umap[:,1].min() - 2, umap[:,1].max() + 2], [z_center, z_center], c = 'k', lw = 5)\n",
    "ax.plot([umap[:,0].min() - 2, umap[:,0].max() + 2], [y_center, y_center], [z_center, z_center], c = 'k', lw = 5)\n",
    "    \n",
    "ax.axis('off')\n",
    "\n",
    "plt.savefig(f'figs/{i:003}', dpi = 100, facecolor = '#fbf8f1')\n",
    "\n",
    "plt.show()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "dc7c4f98-21d8-4b06-a6c0-7e017306d8df",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "754c1674-424d-4572-9a03-dbb6814e977f",
   "metadata": {},
   "outputs": [],
   "source": [
    "for i in range(0, 360, 2):\n",
    "    fig = plt.figure(figsize = (10,10))\n",
    "    ax = fig.add_subplot(projection = '3d')\n",
    "    \n",
    "    ax.scatter(umap[:,0],umap[:,1],umap[:,2], \n",
    "               c = adata_downsampled.obs['leiden_1.0'].astype('int').map(color))\n",
    "    #position\n",
    "    x_center = (umap[:,0].max() + umap[:,0].min())/2\n",
    "    y_center = (umap[:,1].max() + umap[:,1].min())/2\n",
    "    z_center = (umap[:,2].max() + umap[:,2].min())/2\n",
    "\n",
    "    #lines\n",
    "    ax.plot([x_center,x_center], [y_center, y_center], [umap[:,2].min() - 2, umap[:,2].max() + 2], c = 'k', lw = 5)\n",
    "    ax.plot([x_center,x_center], [umap[:,1].min() - 2, umap[:,1].max() + 2], [z_center, z_center], c = 'k', lw = 5)\n",
    "    ax.plot([umap[:,0].min() - 2, umap[:,0].max() + 2], [y_center, y_center], [z_center, z_center], c = 'k', lw = 5)\n",
    "    \n",
    "    ax.view_init(20, i)\n",
    "    \n",
    "    ax.axis('off')\n",
    "\n",
    "    plt.savefig(f'figs/{i:003}', dpi = 100, facecolor = '#fbf8f1')\n",
    "\n",
    "    #plt.legend()\n",
    "    plt.show()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "657d62cf-d0b0-482d-8a55-2ac8c1f7988a",
   "metadata": {},
   "outputs": [],
   "source": [
    "!magick -delay 5 figs/*.png MWBs_leiden_1.0.gif"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3 (ipykernel)",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.13.12"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
